### ISSUE TITLE
Adding a method to a Java annotation breaks tasty-compatibility

### ISSUE THREAD

**Issue by Florian3k** [scala/scala3#19951]
## Compiler version

3.4.0, 3.3.3 (also 3.2.0, 3.1.0, 3.0.0)

## Minimized code
Run on jvm newer than 8:

File `test.scala`:
```scala
object Foo {
  @Deprecated
  def foo(): Unit = ???
}
```
`scalac test.scala -release:8 -d Foo.jar`
`scalac -decompile Foo.jar`
## Output
Fails with
```scala
-- Error: test.scala:2:4 -----------------------------------------------------
2 |    @Deprecated
  |    ^^^^^^^^^^^
  |wrong number of arguments at <no phase> for (since: String, forRemoval: Boolean): Deprecated: (Deprecated#<init> : (since: String, forRemoval: Boolean): Deprecated), expected: 2, found: 0
1 error found
```

## Expectation
Something like that:
```scala
/** Decompiled from Foo.jar(Foo.tasty) */
@scala.annotation.internal.SourceFile("test.scala") object Foo {
  @java.lang.Deprecated def foo(): scala.Unit = scala.Predef.???
}
```

**Comment by smarter** [comment]
There's a few things going on here:
- When we read a Java annotation from a classfile, we pretend it has a constructor whose arguments are all the methods defined in the annotation: https://github.com/scala/scala3/blob/e2c456f08a35d2cae5727c381c63a9e16936b583/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala#L1036-L1054
- When we write `@foo(args)` we create an `Annotation` structure which wraps a `tree` created by typechecking `new foo(args)`
- When tasty pickles an annotation, it pickles that tree.

Adding a new method with a default value to an annotation definition is binary-compatible (https://stackoverflow.com/a/31721504), but it breaks tasty since it changes the annotation constructor,. This is a design flaw in tasty and is going to be hard to fix properly.

Note that if we start from something like:
```java
public @interface Annot {
  String a() default "";
}
```
```scala
@Annot("hi") def foo = 1
```
And we then add an extra parameter:
```java
public @interface Annot {
  String a() default "";
  String b() default "";
}
```
... then when unpickling the annotation on `def foo`, we don't have enough information to even know whether the argument `"hi"` corresponds to the parameter `a` or `b`. If the typechecker desugared the user call to:
```scala
@Annot(a = "hi") def foo = 1
```
Then we'd have the necessary information in Tasty since we'll see a NAMEDARG node in that case.


<hr/>

So one plan would be to:
- Change the typechecker to use named arguments for annotation constructors
- Special-case unpickling from tasty for Java annotation constructors: fill-up missing arguments with nme.WILDCARD (see https://github.com/scala/scala3/commit/85cd1cf3f2e549d02ca74845672a8e585b3fc727).
  - For old tasty where we don't have named arguments, we might still be able to fill-up missing arguments based on type alone if there is no ambiguity, otherwise emit an error.

Alternatively, we could envision some more drastic changes in tasty, but that would prevent us from backporting the fix to 3.3.

/cc @sjrd since whatever scheme we come up with needs to be compatible with tasty-query .

**Comment by smarter** [comment]
Another (unlikely) way for the existing code to do the wrong thing would be the case of a java annotation where the author decided to swap the order of the definitions, for example, if I have:
```java
public @interface Annot {
  String a() default "";
  String b() default "";
}
```
```scala
@Annot("hi", "hello") def foo = 1
```
and one day `Annot` is changed to:
```java
public @interface Annot {
  String b() default "";
  String a() default "";
}
```
Then when we unpickle `@Annot("hi", "hello")`, we will assign them the wrong way around. I think we have to live with that issue for existing tasty files. For new ones we have to rely on unpickling NamedArg trees and having the TypeAssigner perform the swapping and insertion of nme.WILDCARD.
